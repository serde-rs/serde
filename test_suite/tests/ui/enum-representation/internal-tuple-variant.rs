use serde_derive::{Deserialize, Serialize};

#[derive(Serialize)]
#[serde(tag = "type")]
enum Serializable {
    /// No errors should be reported, equivalent to Unit representation
    Tuple0(),
    /// No errors should be reported, delegates decision to the inner type
    Tuple1(u8),
    /// Error should be reported
    Tuple2(u8, u8),

    /// No errors should be reported, equivalent to Tuple0 representation
    Tuple1as0S(#[serde(skip_serializing)] u8),
    /// No errors should be reported, equivalent to Tuple1 representation
    Tuple1as0D(#[serde(skip_deserializing)] u8),

    /// No errors should be reported, equivalent to Tuple1 representation
    Tuple2as1S(#[serde(skip_serializing)] u8, u8),
    /// Error should be reported, equivalent to Tuple2 representation
    Tuple2as1D(#[serde(skip_deserializing)] u8, u8),

    /// Error should be reported, equivalent to Tuple2 representation
    Tuple3as2S(#[serde(skip_serializing)] u8, u8, u8),
    /// Error should be reported, equivalent to Tuple3 representation
    Tuple3as2D(#[serde(skip_deserializing)] u8, u8, u8),
}

#[derive(Deserialize)]
#[serde(tag = "type")]
enum Deserializable {
    /// No errors should be reported, equivalent to Unit representation
    Tuple0(),
    /// No errors should be reported, delegates decision to the inner type
    Tuple1(u8),
    /// Error should be reported
    Tuple2(u8, u8),

    /// No errors should be reported, equivalent to Tuple1 representation
    Tuple1as0S(#[serde(skip_serializing)] u8),
    /// No errors should be reported, equivalent to Tuple0 representation
    Tuple1as0D(#[serde(skip_deserializing)] u8),

    /// Error should be reported, equivalent to Tuple2 representation
    Tuple2as1S(#[serde(skip_serializing)] u8, u8),
    /// No errors should be reported, equivalent to Tuple1 representation
    Tuple2as1D(#[serde(skip_deserializing)] u8, u8),

    /// Error should be reported, equivalent to Tuple3 representation
    Tuple3as2S(#[serde(skip_serializing)] u8, u8, u8),
    /// Error should be reported, equivalent to Tuple2 representation
    Tuple3as2D(#[serde(skip_deserializing)] u8, u8, u8),
}

#[derive(Serialize, Deserialize)]
#[serde(tag = "type")]
enum Both {
    /// No errors should be reported, equivalent to Unit representation
    Tuple0(),
    /// No errors should be reported, delegates decision to the inner type
    Tuple1(u8),
    /// Error should be reported
    Tuple2(u8, u8),

    /// Serialize: No errors should be reported, equivalent to Tuple0 representation
    /// Deserialize: No errors should be reported, equivalent to Tuple1 representation
    Tuple1as0S(#[serde(skip_serializing)] u8),
    /// Serialize: No errors should be reported, equivalent to Tuple1 representation
    /// Deserialize: No errors should be reported, equivalent to Tuple0 representation
    Tuple1as0D(#[serde(skip_deserializing)] u8),

    /// Serialize: No errors should be reported, equivalent to Tuple1 representation
    /// Deserialize: Error should be reported, equivalent to Tuple2 representation
    Tuple2as1S(#[serde(skip_serializing)] u8, u8),
    /// Serialize: Error should be reported, equivalent to Tuple2 representation
    /// Deserialize: No errors should be reported, equivalent to Tuple1 representation
    Tuple2as1D(#[serde(skip_deserializing)] u8, u8),

    /// Serialize: Error should be reported, equivalent to Tuple2
    /// Deserialize: Error should be reported, equivalent to Tuple3
    Tuple3as2S(#[serde(skip_serializing)] u8, u8, u8),
    /// Serialize: Error should be reported, equivalent to Tuple3
    /// Deserialize: Error should be reported, equivalent to Tuple2
    Tuple3as2D(#[serde(skip_deserializing)] u8, u8, u8),
}

fn main() {}
