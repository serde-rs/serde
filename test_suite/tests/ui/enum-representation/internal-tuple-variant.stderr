error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:10:5
   |
10 | /     /// Error should be reported
11 | |     Tuple2(u8, u8),
   | |__________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:20:5
   |
20 | /     /// Error should be reported, equivalent to Tuple2 representation
21 | |     Tuple2as1D(#[serde(skip_deserializing)] u8, u8),
   | |___________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:23:5
   |
23 | /     /// Error should be reported, equivalent to Tuple2 representation
24 | |     Tuple3as2S(#[serde(skip_serializing)] u8, u8, u8),
   | |_____________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:25:5
   |
25 | /     /// Error should be reported, equivalent to Tuple3 representation
26 | |     Tuple3as2D(#[serde(skip_deserializing)] u8, u8, u8),
   | |_______________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:36:5
   |
36 | /     /// Error should be reported
37 | |     Tuple2(u8, u8),
   | |__________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:44:5
   |
44 | /     /// Error should be reported, equivalent to Tuple2 representation
45 | |     Tuple2as1S(#[serde(skip_serializing)] u8, u8),
   | |_________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:49:5
   |
49 | /     /// Error should be reported, equivalent to Tuple3 representation
50 | |     Tuple3as2S(#[serde(skip_serializing)] u8, u8, u8),
   | |_____________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:51:5
   |
51 | /     /// Error should be reported, equivalent to Tuple2 representation
52 | |     Tuple3as2D(#[serde(skip_deserializing)] u8, u8, u8),
   | |_______________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:62:5
   |
62 | /     /// Error should be reported
63 | |     Tuple2(u8, u8),
   | |__________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:75:5
   |
75 | /     /// Serialize: Error should be reported, equivalent to Tuple2 representation
76 | |     /// Deserialize: No errors should be reported, equivalent to Tuple1 representation
77 | |     Tuple2as1D(#[serde(skip_deserializing)] u8, u8),
   | |___________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:79:5
   |
79 | /     /// Serialize: Error should be reported, equivalent to Tuple2
80 | |     /// Deserialize: Error should be reported, equivalent to Tuple3
81 | |     Tuple3as2S(#[serde(skip_serializing)] u8, u8, u8),
   | |_____________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:82:5
   |
82 | /     /// Serialize: Error should be reported, equivalent to Tuple3
83 | |     /// Deserialize: Error should be reported, equivalent to Tuple2
84 | |     Tuple3as2D(#[serde(skip_deserializing)] u8, u8, u8),
   | |_______________________________________________________^

error: #[serde(tag = "...")] cannot be used with tuple variants
  --> tests/ui/enum-representation/internal-tuple-variant.rs:72:5
   |
72 | /     /// Serialize: No errors should be reported, equivalent to Tuple1 representation
73 | |     /// Deserialize: Error should be reported, equivalent to Tuple2 representation
74 | |     Tuple2as1S(#[serde(skip_serializing)] u8, u8),
   | |_________________________________________________^
